### PR TITLE
Add ESLint starter workflow

### PR DESCRIPTION
**Pull request by yongyan-gh**
## Pre-requisites

- [x] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least priviledge [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully. 

**For _CI_ workflows, the workflow:**

- [ ] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [x] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [x] `name`: Name of the Code Scanning integration.
  - [ ] `organization`: Name of the organization producing the Code Scanning integration.
  - [x] `description`: Short description of the Code Scanning integration.
  - [x] `categories`: Array of languages supported by the Code Scanning integration.
  - [x] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [x] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [x] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [x] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [x] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

File: code-scanning/eslint.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+jobs:
+  eslint:
+    name: Run eslint scanning
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      security-events: write
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
```
Comment (eddynaka) [inline code-scanning/eslint.yml:26]: should we start using v3? I saw it was released a few days ago

File: code-scanning/properties/eslint.properties.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "ESLint",
+    "creator": "ESLint",
+    "description": "A tool for identifying and reporting the problems found in ECMAScript/JavaScript code.",
+    "iconName": "eslint",
+    "categories": [
+        "Code Scanning",
+        "javascript",
```
Comment (eddynaka) [inline code-scanning/properties/eslint.properties.json:8]: should we add typescript?

File: code-scanning/eslint.yml (hunk 1 above)
Comment (yongyan-gh) [reply code-scanning/eslint.yml:26]: The checkout v3 readme says:
> - Updated to the node16 runtime by default
>   - This requires a minimum [Actions Runner](https://github.com/actions/runner/releases/tag/v2.285.0) version of v2.285.0 to run, which is by default available in GHES 3.4 or later.

Any possibility the server has an old version runner and cannot run v3 checkout?

File: code-scanning/properties/eslint.properties.json (hunk 2 above)
Comment (yongyan-gh) [reply code-scanning/properties/eslint.properties.json:8]: Added

**Comment by NickLiffen** [comment]
Just got around to testing this out 👍 sorry only just got back from holiday. 

The only thought I had is around this:

Screenshot 2022-03-24 at 12 22 51

Right now there is no helper text going into this, have we thought about adding helper text into here? 

There is a lot of great information here: https://eslint.org/docs/rules/no-unused-vars under the rules section? Could we put a link or something to the page with the rule information?

**Comment by josepalafox** [comment]
Cc @eddynaka

**Comment by eddynaka** [comment]
Hi @NickLiffen and @josepalafox ,

talking with @yongyan-gh , he mentioned that the SARIF we generate contains a field called [rule.HelpUri](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317848) which contains the link from eslint or from the custom provider.

I think the best approach here is to enable that in GHAS.

@yongyan-gh , can you attach a SARIF file generated by the last formatter, pls?

**Comment by yongyan-gh** [comment]
@NickLiffen thanks very much for your input!

The SARIF output of ESLint contains the doc url in its `rule.helpUri' property. E.g.
```
          "rules": [
            {
              "id": "no-undef",
              "shortDescription": {
                "text": "disallow the use of undeclared variables unless mentioned in `/*global */` comments"
              },
              "helpUri": "https://eslint.org/docs/rules/no-undef",
              "properties": {}
            },
```
Which converted from ESLint results metadata:
```
            "no-undef": {
                "type": "problem",
                "docs": {
                    "description": "disallow the use of undeclared variables unless mentioned in `/*global */` comments",
                    "recommended": true,
                    "url": "https://eslint.org/docs/rules/no-undef"
                },
```
Not sure which property can be used to render the rule id with a hyperlink to the rule doc.
Attached an example ESLint SARIF log file: [eslint.zip](https://github.com/actions/starter-workflows/files/8344226/eslint.zip)

File: code-scanning/eslint.yml (hunk 1 above)
Comment (bishal-pdMSFT) [reply code-scanning/eslint.yml:26]: Please use v3 version. New server releases have new runner and hence won't be a problem.

File: code-scanning/properties/eslint.properties.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "ESLint",
+    "creator": "ESLint",
```
Comment (bishal-pdMSFT) [inline code-scanning/properties/eslint.properties.json:3]: Who is the POC for this?

Comment (josepalafox) [reply code-scanning/properties/eslint.properties.json:3]: Me

File: code-scanning/eslint.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Run ESLint
+        run: npx eslint .
+          --config .eslintrc.js
+          --ext .js,.jsx,.ts,.tsx
+          --format @microsoft/eslint-formatter-sarif 
+          --output-file eslint-results.sarif
+        continue-on-error: true
+
+      - name: Upload analysis results to GitHub
+        uses: github/codeql-action/upload-sarif@v1
```
Comment (bishal-pdMSFT) [inline code-scanning/eslint.yml:42]: I think this one as well should be v2.

File: code-scanning/eslint.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# This workflow uses actions that are not certified by GitHub.
```
Comment (bishal-pdMSFT) [inline code-scanning/eslint.yml:1]: Please add a comment on what this workflow does

**Review by bishal-pdMSFT — COMMENTED** [review]
Mostly looks good to me other than few minor comments.
@josepalafox would you also like to sign off on this?

File: code-scanning/eslint.yml (hunk 4 above)
Comment (yongyan-gh) [reply code-scanning/eslint.yml]: updated to v2

File: code-scanning/eslint.yml (hunk 5 above)
Comment (yongyan-gh) [reply code-scanning/eslint.yml:1]: added comments

File: code-scanning/eslint.yml (hunk 1 above)
Comment (eddynaka) [reply code-scanning/eslint.yml:30]: @yongyan-gh , you missed this one

File: code-scanning/properties/eslint.properties.json (hunk 3 above)
Comment (bishal-pdMSFT) [reply code-scanning/properties/eslint.properties.json:3]: @yongyan-gh please remove this line as @josepalafox's team from GitHub would be the owners.

Comment (yongyan-gh) [reply code-scanning/properties/eslint.properties.json]: removed the creator property

File: code-scanning/eslint.yml (hunk 1 above)
Comment (yongyan-gh) [reply code-scanning/eslint.yml]: updated version to v3

File: code-scanning/properties/eslint.properties.json (hunk 2 above)
Comment (anuragc617) [reply code-scanning/properties/eslint.properties.json:7]: For consistency please follow the Pascal naming convention as used in other workflows: "JavaScript", "TypeScript", "EcmaScript"

File: code-scanning/properties/eslint.properties.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "ESLint",
+    "description": "A tool for identifying and reporting the problems found in ECMAScript/JavaScript code.",
+    "iconName": "eslint",
+    "categories": [
+        "Code Scanning",
+        "javascript",
+        "ecmascript",
```
Comment (anuragc617) [inline code-scanning/properties/eslint.properties.json:8]: Is there way do detect EcmaScript? https://github.com/github/linguist/blob/master/lib/linguist/languages.yml doesn;t have logic for it?

Comment (yongyan-gh) [reply code-scanning/properties/eslint.properties.json]: My understanding is that the ECMAScript is spec/standard for scripting languages like JavaScript. I think the ESLint inspects the source code by checking the ECMAScript specification. Should it be added in the categories?

**Review by bishal-pdMSFT — APPROVED** [review]

**Pull request merged**